plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergStatistics.java

```diff
@@ -0,0 +1,325 @@
+                List<Types.NestedField> columns,
+                TypeManager typeManager)
+        {
+            this.idToTypeMapping = requireNonNull(idToTypeMapping, "idToTypeMapping is null");
+            this.columns = requireNonNull(columns, "columns is null");
+            this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        }
+
+        public void registerDataFile(DataFile dataFile, PartitionSpec partitionSpec)
+        {
+            fileCount++;
+            recordCount += dataFile.recordCount();
+            size += dataFile.fileSizeInBytes();
+
+            Map<Integer, Long> newColumnSizes = dataFile.columnSizes();
+            if (newColumnSizes != null) {
+                for (Types.NestedField column : columns) {
+                    int id = column.fieldId();
+                    Long addedSize = newColumnSizes.get(id);
+                    if (addedSize != null) {
+                        columnSizes.put(id, addedSize + columnSizes.getOrDefault(id, 0L));
+                    }
+                }
+            }
+
+            Set<Integer> partitionFieldIds = partitionSpec.fields().stream()
+                    .filter(field -> field.transform().isIdentity())
+                    .map(PartitionField::sourceId)
+                    .collect(toImmutableSet());
+            Map<Integer, Optional<String>> partitionValues = getPartitionKeys(dataFile.partition(), partitionSpec);
+            Map<Integer, Object> lowerBounds = convertBounds(idToTypeMapping, dataFile.lowerBounds());
+            Map<Integer, Object> upperBounds = convertBounds(idToTypeMapping, dataFile.upperBounds());
+            for (Types.NestedField column : columns) {
+                int id = column.fieldId();
+                io.trino.spi.type.Type trinoType = toTrinoType(column.type(), typeManager);
+                if (partitionFieldIds.contains(id)) {
+                    Optional<String> partitionValue = partitionValues.get(id);
+                    if (partitionValue.isPresent()) {
+                        Type type = idToTypeMapping.get(id);
+                        Object trinoValue = deserializePartitionValue(toTrinoType(type, typeManager), partitionValue.get(), column.name());
+                        // Update min/max stats but there are no null values to count
+                        updateMinMaxStats(
+                                id,
+                                trinoType,
+                                trinoValue,
+                                trinoValue,
+                                Optional.of(0L),
+                                dataFile.recordCount());
+                        updateNullCountStats(id, Optional.of(0L));
+                    }
+                    else {
+                        // Update null counts, but do not clear min/max
+                        updateNullCountStats(id, Optional.of(dataFile.recordCount()));
+                    }
```

Comment:
I see what you're getting at. I kinda like the separation as it is because there's a "this partition value is non-null" block, and a "this partition value is null" block.